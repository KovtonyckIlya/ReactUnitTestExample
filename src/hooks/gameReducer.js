import { useReducer} from 'react';
import gameReducer from '../reducer/gameReducer';

export default function useGameReducer () {
  const [state, dispatch] = useReducer(gameReducer, {
    squares: Array(9).fill(null),
    isXNext: true,
    move: null,
    gameId: null,
    roomName: null,
    username: '',
    allPlayers: [],
    otherUser: [], //changed
    multiplayer: false, //changed
    gameFull: false, // changed
    clientID: null,
    chat: [],
    gameOver: false,
    response: {},
    rematch: false,
    error: false,
    gamesChanged: false,
    topBarResponse: {},
    host: false //changed
  }); 
  return [state, dispatch];
};