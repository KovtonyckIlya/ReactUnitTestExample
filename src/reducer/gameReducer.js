export default function gameReducer(state, action) {
  const {
    squares,
    isXNext,
    firstMove,
    move,
    gameId,
    roomName,
    username,
    otherUser,
    allPlayers,
    multiplayer,
    gameFull,
    clientID,
    chat,
    gameOver,
    response,
    rematch,
    gamesChanged,
    topBarResponse,
    host
  } = state;

  switch (action.type) {
    case 'SELECT_SQUARE': {
      const currentSquares = squares.slice();
      currentSquares[action.index] = isXNext ? 'X' : 'O';
      return {
        ...state,
        squares: currentSquares,
        isXNext: !isXNext
      }
    }
    case 'UPDATE_SQUARES': {
      return {
        ...state,
        squares: action.squares,
        isXNext: action.isXNext
      }
    }
    case 'RESET_SQUARE': {
      return {
        ...state,
        squares: Array(9).fill(null),
        isXNext: true
      }
    }
    case 'createGame': {
      return {
        ...state,
        gameId: action.data.gameId,
        roomName: action.data.roomName,
        move: action.data.move,
        username: action.data.username,
        clientID: action.data.clientID,
        multiplayer: true,
        host: true
      };
    } 
    case 'joinGame': {
      return {
        ...state,
        gameId: action.data.gameId,
        move: action.data.move,
        roomName: action.data.roomName,
        username: action.data.username,
        clientID: action.data.clientID,
        multiplayer: true,
        host: false
      };
    }
    case 'notifyAllUsers': {
      const you = {username: username, move: move, clientID: clientID, host: host};
      const otherUser = action.data.users.filter(user => user.move !== move);
      const players = [you].concat(otherUser[0]);

      return {
        ...state,
        otherUser: otherUser,
        allPlayers: players,
        isXNext: action.data.users[0].move === 'X' ? true : false,
        firstMove: action.data.users[0].move,
        response: {
          type: 'userJoined',
          userThatJoined: action.data.userThatJoined
        },
        gameFull: true
      };
    }
    case "updateSquares": {
      const currentSquares = squares.slice();
      currentSquares[action.data.index] = isXNext ? 'X' : 'O';
      return {
        ...state,
        squares: currentSquares,
        isXNext: !isXNext
      };
    }
    case "messageDelivered": {
      return {
        ...state,
        delivered: true
      };
    }
    case "newMessages" : {
      const newMessages = chat.concat({
        message: action.data.message,
        clientID: action.data.clientID,
        username: action.data.username
      })
      return {
        ...state,
        chat: newMessages,
        response: {
          type: 'latestMessage',
          username: action.data.username,
          message: action.data.message
        }
      }
    }
    case "UPDATE_USERNAME": {
      return {
        ...state,
        username: action.username
      };
    }
    case "leaveGame": {
      return {
        ...state,
        squares: Array(9).fill(null),
        isXNext: true,
        move: null,
        gameId: null,
        roomName: null,
        rematch: false,
        gameOver: false,
        allPlayers: [],
        otherUser: [],
        host: false,
        multiplayer: false,
        gameFull: false,
        chat: []
      };
    }
    case "kicked": {
      return {
        ...state,
        squares: Array(9).fill(null),
        isXNext: true,
        move: null,
        gameId: null,
        roomName: null,
        rematch: false,
        gameOver: false,
        allPlayers: [],
        otherUser: [],
        host: false,
        multiplayer: false,
        gameFull: false,
        chat: [],
        topBarResponse: {
          type: 'MESSAGE',
          message: 'You got kicked lol'
        }
      }
    }
    case "playerwaskicked": {
      return {
        ...state,
        squares: Array(9).fill(null),
        rematch: false,
        gameOver: false,
        allPlayers: [],
        otherUser: [],
        gameFull: false,
        host: true,
        chat: []
      }
    }
    case "notifyUserLeft": {
      return {
        ...state,
        squares: Array(9).fill(null),
        rematch: false,
        gameOver: false,
        allPlayers: [],
        otherUser: [],
        gameFull: false,
        host: true,
        chat: [],
        response: {
          type: 'userLeft',
          username: action.data.username
        }
      }
    }
    case "GAME_OVER": {
      return {
        ...state,
        gameOver: true
      }
    }
    case "CLEAR_RESPONSE": {
      return {
        ...state,
        response: {}
      }
    }
    case "removeRematchButton": {
      return {
        ...state,
        rematch: true
      }
    }
    case "rematch": {
      return {
        ...state,
        response: {
          type: 'rematch',
          username: action.data.username
        }
      }
    }
    case "rematchStart" : {
      return {
        ...state,
        response: {
          type: 'rematchStarted'
        },
        rematch: false,
        squares: Array(9).fill(null),
        isXNext: firstMove === 'X' ? false : true,
        firstMove: firstMove === 'X' ? 'O' : 'X',
        gameOver: false
      }
    }
    case "error" : {
      return {
        ...state,
        topBarResponse: {
          type: "ERROR",
          message: action.data.message 
        }
      }
    }
    case "playerDisconnect" : {
      return {
        ...state,
        gameFull: false,
        allPlayers: [],
        otherUser: [],
        response: {
          type: 'playerDisconnect',
          playerName: action.data.player.username
        },
        rematch: false,
        squares: Array(9).fill(null),
        isXNext: true,
        gameOver: false,
        chat: [],
      }
    }
    case "gamesChanged" : {
      return {
        ...state,
        gamesChanged: !gamesChanged,
        topBarResponse: {
          type: 'gamesUpdated'
        }
      };
    }
    case "RESET_STATE" : {
      return {
        ...state,
        squares: Array(9).fill(null),
        isXNext: true,
        move: null,
        gameId: null,
        roomName: null,
        allPlayers: [],
        otherUser: [],
        multiplayer: false,
        gameFull: false,
        clientID: null,
        chat: [],
        gameOver: false,
        response: {},
        rematch: false,
        error: false,
        gamesChanged: false
      }
    }
    case "TOP_BAR_RESPONSE" : {
      return {
        ...state,
        topBarResponse: {
          type: action.data.type,
          message: action.data.message
        }
      }
    }
    case "CLEAR_TOP_BAR_RESPONSE" : {
      return {
        ...state,
        topBarResponse: {}
      }
    }
  };
};