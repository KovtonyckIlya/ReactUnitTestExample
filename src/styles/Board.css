.boardContainer {
  display: flex;
  flex-direction: column;
}

.boardRow {
  display: flex;
  height: 80px;
  padding: 0px;
  margin: 0px;
}

.status {
  padding-top: 5px;
  padding-bottom: 5px;
  color: #444;
}

.resetButton {
  padding: 10px;
  background: #03a9f4;
  border-radius: 5px;
  color: white;
  font-weight: bold;
  cursor: pointer;
}