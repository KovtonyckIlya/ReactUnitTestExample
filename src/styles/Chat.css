.chatContainer {
  width: 100%;
  background: white;
  border: 2px solid #ccc;
  border-radius: 5px;
  position: relative;
}

.chatContainerMobile {
  background: white;
  z-index: 999;
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  width: 100%;
}

.anotherContainer {
  display: flex;
  flex-direction: column;
  width: 100%;
  z-index: 999;
  background: white;
  bottom: 0;
  position: absolute;
}

.anotherContainerMobile {
  display: flex;
  flex-direction: column;
  width: 100%;
  z-index: 999;
  background: white;
  bottom: 0;
  position: fixed;
}

.subContainer{
  height: 40px;
  display: flex;
  align-items: center;
}

.closeButton {
  font-weight: bold;
  padding: 10px;
  font-size: 20px;
  color: #03a9f4; 
  cursor: pointer;
}

.sendMessage {
  font-weight: bold;
  padding: 10px;
  font-size: 18px;
  color: #03a9f4;
  cursor: pointer;
}

.chatInput {
  flex: 1;
  border: none;
  border-radius: 0;
  outline: none;
  box-shadow: none;
  font-size: 16px;
  padding: 5px;
  background: white;
}

.line {
  width: 90%;
  align-self: center;
  border-bottom: 2px solid #ccc;
}

.mainMessageContainer {
  overflow-y: scroll;
  height: 482px;
  padding-top: 60px;
  padding-bottom: 60px;
}

.mainMessageContainerMobile {
  padding-top: 60px;
  padding-bottom: 60px;
  background: white;
}

.scroll {
  overflow-y: scroll;
  -webkit-overflow-scrolling: touch;
  border-bottom-right-radius: 10px;
  border-bottom-left-radius: 10px;
}

.messageContainerLeft {
  display: flex;
  justify-content: flex-start;
}

.messageContainerRight {
  display: flex;
  justify-content: flex-end;
}

.chatLeft, .chatRight {
  display: inline-block;
  padding: 10px;
  font-size: 20px;
  max-width: 70%;
  margin-top: 2px;
  border-radius: 10px;
}

.chatLeft {
  background: #ccc;
  color: #444;
  margin-left: 15px;
}

.chatRight {
  background: #03a9f4;
  color: white;
  margin-right: 15px;
}

.playerNameContainer {
  position: absolute;
  height: 50px;
  background: #eaeded;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.playerNameContainerMobile {
  position: fixed;
  height: 50px;
  background: #eaeded;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.playerName {
  color: #444;
  font-size: 20px;
}

.closeButtonTop {
  font-weight: bold;
  padding: 10px;
  font-size: 16px;
  color: #03a9f4; 
  cursor: pointer;
  position: absolute;
  left: 5px;
}