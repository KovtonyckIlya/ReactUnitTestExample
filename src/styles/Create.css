.mainContainer {
  width: 100%;
  box-sizing: border-box;
  align-items: center;
  font-family: sans-serif;
  display: flex;
  flex-direction: column;
  border-radius: 5px;
  background: #03a9f4;
  width: 100%;
  position: relative;
}

.menuContainer {
  width: 100%;
  padding-bottom: 15px;
}

.container {
  padding-left: 15px;
  padding-bottom: 15px;
  padding-right: 15px;
  width: 100%;
}

.optionsContainer {
  display: flex;
  flex-direction: column;
  background: white;
  padding: 15px;
  border-radius: 5px;
}

.label {
  color: #444;
  font-size: 20px;
  margin-top: 10px;
}

.marginTop {
  margin-top: 10px;
}

.button {
  color: white;
  background: #03a9f4;
  border: none;
  outline: none;
  padding: 10px;
  font-weight: bold;
  border-radius: 5px;
  font-size: 20px;
  margin-top: 10px;
  cursor: pointer;
}

.moveLabel {
  font-size: 20px;
  font-weight: bold;
  color: #444;
  margin-left: 5px;
}

.input {
  margin-top: 10px;
  font-size: 16px;
}

.error {
  color: red;
  padding-top: 2px;
  padding-bottom: 2px;
}