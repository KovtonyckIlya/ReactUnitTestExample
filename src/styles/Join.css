.container {
  width: 100%;
  box-sizing: border-box;
  align-items: center;
  font-family: sans-serif;
  display: flex;
  flex-direction: column;
  background: #03a9f4;
  border-radius: 5px;
  position: relative;
}

.searchingContainer {
  width: 100%;
  left: 0;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  padding-bottom: 15px;
  background: #03a9f4;
  z-index: 999;
  transition: top 0.5s;
  position: fixed;
}

.searchingContainerDesktop {
  width: 500px;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
  padding-bottom: 15px;
  background: #03a9f4;
  z-index: 999;
  transition: top 0.5s;
  position: fixed;
}

.subContainer {
  width: 100%;
  padding-left: 15px;
  padding-bottom: 15px;
  padding-right: 15px;
  border-radius: 5px;
  transition: padding 0.5s;
  display: flex;
  justify-content: center;
  width: 100%;
}

.bottomContainer {
  width: 100%;
}

.noGames {
  color: white;
  display: flex;
  justify-content: center;
  padding-top: 20px;
}

.gamesContainer, .noGames {
  background: #03a9f4;
}

.loader {
  color: white;
}

.loaderTwoContainer {
  display: flex;
  justify-content: center;
  transition: margin 0.5s;
}