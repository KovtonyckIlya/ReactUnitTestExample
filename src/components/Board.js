import React,{ useEffect, useState }from 'react';
import Square from './Square';
import styles from 'styles/Board.css';
import calculateWinner from '../helpers/calculateWinner';
import chunkArray from '../helpers/chunkArray';

function isPlayerNext(state) {
  const {isXNext,allPlayers} = state;
  const nextMove = (isXNext ? 'X' : 'O');
  console.log(allPlayers)
  const nextPlayer = allPlayers
    .filter(user => user.move === nextMove)
    .reduce((acc, current) => ({...acc, ...current}), {});

  return nextPlayer;
};

function gameStatus(state) {
  let {squares, isXNext, allPlayers, multiplayer, gameFull} = state;
  let winner = calculateWinner(squares);
  let nextMove = (isXNext ? 'X' : 'O');
  let status;
  let gameOver;

  if (winner) {
    if (multiplayer) {
      const playerThatWon = allPlayers
        .filter(user => user.move === winner)
        .reduce((acc, current) => ({...acc, ...current}), {});

      status = playerThatWon.username + ' is the winner!';
    } else {
      status = 'Winner: ' + winner;
    };
    gameOver = true;
  } else if (!winner && squares.every(item => item !== null)) {
    status = 'Game was a tie';
    gameOver = true;
  } else if (multiplayer && !gameFull) {
    status = 'Waiting for second player to join.'
  } else if (multiplayer && gameFull) {
    const nextPlayer = isPlayerNext(state);
    status = 'Next player: ' + nextPlayer.username;
  } else {
    status = 'Next player: ' + nextMove;
  };
  return {
    status,
    gameOver
  };
};

function Board({state, dispatch, sendMessage}) {
  const { squares, gameFull, multiplayer, clientID } = state;
  const status = gameStatus(state);
  const [timeOut, storeTimeOut] = useState(null);
  const [notYourTurn, setNotYourTurn] = useState(false);
  const chunkedArray = chunkArray(squares, 3);

  useEffect(() => {
    if (status.gameOver === true) {
      dispatch({type: 'GAME_OVER'})
    };
  }, [status.gameOver])

  useEffect(() => {
    if (notYourTurn !== false) {
      clearTimeout(timeOut)
      const timeout = setTimeout(() => {
        setNotYourTurn(false)
      }, 1000)
      storeTimeOut(timeout)
    }
    return () => {
      clearTimeout(timeOut)
    }
  }, [notYourTurn])

  function handleClick(i) {
    if (calculateWinner(squares) || squares[i]) {
      return;
    };
    if (multiplayer) {
      if (!gameFull) {
        return;
      }
      const nextPlayer = isPlayerNext(state);
      if (clientID !== nextPlayer.clientID) {
        setNotYourTurn(true);
        return;
      }
      const msg = {
        type: "updateSquares",
        index: i,
        gameId: state.gameId
      };
      sendMessage(msg);
    } else {
      dispatch({type: 'SELECT_SQUARE', index: i});
    };
  };

  function renderSquare(i) {
    return (
      <Square
        key={i}
        value={squares[i]}
        onClick={() => handleClick(i)}
      />
    );
  };

  return (
    <div className={styles.boardContainer}>
      <div id='status' className={styles.status}>{notYourTurn ? "It's not your turn." : status.status}</div>
      {chunkedArray.map((array, i) => {
        return (
          <div key={i} className={styles.boardRow}>
            {array.map((_, index) => {
              const chunkIndex = i > 0 ? ((i * 3) + index) : index;
              return renderSquare(chunkIndex);
            })}
          </div>
        );
      })}
    </div>
  );
};

export default Board;