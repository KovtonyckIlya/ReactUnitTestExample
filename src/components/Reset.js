import React from 'react';
import styles from '../styles/Board.css';

export default function Reset ({resetBoard}) {
  return (
    <div id='resetButton' className={styles.resetButton} onClick={resetBoard}>Restart Game</div>
  );
};