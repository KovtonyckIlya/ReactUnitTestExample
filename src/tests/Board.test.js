import React, { useReducer } from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import Board from '../components/Board';
import Reset from '../components/Reset';
import renderer from 'react-test-renderer';
import styles from 'styles/Board.css';
import squareStyles from 'styles/Square.css';
import gameReducer from '../reducer/gameReducer';

let container;

//If test does require dispatch then i pass in a real dispatch from the reducer
function BoardWithReducer () {
  const [state, dispatch] = useReducer(gameReducer, {
    squares: Array(9).fill(null),
    isXNext: true,
    multiplayer: false
  });
  return <Board state={state} dispatch={dispatch}/>
};

//Im testing these two together because reset component resets the squares in the board component they are closely related.
function BoardAndResetWithReducer () {
  const [state, dispatch] = useReducer(gameReducer, {
    squares: Array(9).fill(null),
    isXNext: true,
    multiplayer: false
  });
  //Im not testing this function so im passing a mock
  const resetHistory = jest.fn();
  return (
    <div>
      <Board state={state} dispatch={dispatch}/>
      <Reset resetHistory={resetHistory} dispatch={dispatch}/>
    </div>
  );
};
//If test does not require dispatch is decided to use this
function BoardWithMock () {
  const state = {squares: Array(9).fill(null), isXNext: false};
  const dispatch = jest.fn();
  return <Board state={state} dispatch={dispatch} />
};

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

it('Renders without crashing', () => {
  ReactDOM.render(<BoardWithMock />, container);
});

it('Renders correctly', () => {
  const tree = renderer.create(<BoardWithMock />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render 5 divs', () => {
  act(() => {
    ReactDOM.render(<BoardWithMock />, container);
  });
  const divs = container.querySelectorAll('div');
  expect(divs.length).toEqual(5);
});

it('Each row should have 3 columns.', () => {
  act(() => {
    ReactDOM.render(<BoardWithMock />, container);
  });
  const boardRows = container.querySelectorAll('.' + styles.boardRow);
  boardRows.forEach(row => {
    expect(row.children.length).toEqual(3);
  });
});

it('Board updates with X and O when user clicks on it and renders proper status.', () => {
  act(() => {
    ReactDOM.render(<BoardWithReducer />, container);
  });
  const squares = container.querySelectorAll('button');
  const status = container.querySelector('#status');
  expect(squares[0].textContent).toBe('');
  act(() => {
    squares[0].dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  expect(status.innerHTML).toBe('Next player: O');
  expect(squares[0].textContent).toBe('X');
  expect(squares[1].textContent).toBe('');
  act(() => {
    squares[1].dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  expect(status.innerHTML).toBe('Next player: X');
  expect(squares[1].textContent).toBe('O');
});

it('Should still only render 5 divs if all squares are selected and the game was a tie.', () => {
  act(() => {
    ReactDOM.render(<BoardWithReducer />, container);
  });
  const squares = container.querySelectorAll('button');
  const orderOfClicksByIndex = [0,1,2,4,3,5,7,6,8];
  orderOfClicksByIndex.forEach(index => {
    act(() => {
      squares[index].dispatchEvent(new MouseEvent('click', {bubbles: true}));
    });
  });
  const divs = container.querySelectorAll('div');
  expect(divs.length).toEqual(5);
});

it('Clicking should only change the column that was clicked, dont change the numbers on chunkIndex Variable.', () => {
  act(() => {
    ReactDOM.render(<BoardWithReducer />, container);
  });
  const squares = container.querySelectorAll('.' + squareStyles.square);
  act(() => {
    squares[2].dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  squares.forEach((_, index) => {
    if (index !== 2) {
      expect(squares[index].textContent).toBe('');
    }
  })
});

it('Should display winner when someone wins with their symbol', () => {
  act(() => {
    ReactDOM.render(<BoardWithReducer />, container);
  });
  const squares = container.querySelectorAll('button');
  const orderOfClicksforXtoWin = [0,1,4,2,8];
  orderOfClicksforXtoWin.forEach(index => {
    act(() => {
      squares[index].dispatchEvent(new MouseEvent('click', {bubbles: true}));
    });
  });
  const status = container.querySelector('#status');
  expect(status.innerHTML).toBe('Winner: X');
});

it('If game was a tie status should be Game was a tie.', () => {
  act(() => {
    ReactDOM.render(<BoardWithReducer />, container);
  });
  const squares = container.querySelectorAll('button');
  const orderOfClicksByIndex = [0,1,2,4,3,5,7,6,8];
  const status = container.querySelector('#status');
  orderOfClicksByIndex.forEach(index => {
    act(() => {
      squares[index].dispatchEvent(new MouseEvent('click', {bubbles: true}));
    });
  });
  expect(status.innerHTML).toBe('Game was a tie');
});

it('When you click reset button, the board resets', () => {
  act(() => {
    ReactDOM.render(<BoardAndResetWithReducer />, container);
  });
  const squares = container.querySelectorAll('.' + squareStyles.square);
  const resetButton = container.querySelector('#resetButton');
  const status = container.querySelector('#status');
  act(() => {
    resetButton.dispatchEvent(new MouseEvent('click', {bubbles: true}));
  });
  squares.forEach((_, index) => {
    expect(squares[index].textContent).toBe('');
  });
  expect(status.innerHTML).toBe('Next player: X');
});