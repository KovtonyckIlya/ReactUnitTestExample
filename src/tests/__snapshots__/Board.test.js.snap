// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  className="boardContainer"
>
  <div
    className="status"
    id="status"
  >
    Next player: O
  </div>
  <div
    className="boardRow"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="boardRow"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
  <div
    className="boardRow"
  >
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
    <button
      className="square"
      onClick={[Function]}
    />
  </div>
</div>
`;
